src/recon_buildblock/LogcoshPrior.cxx

```diff
@@ -0,0 +1,609 @@
+              // psi'(t)/t = tanh/t
+              elemT temp = fabs(current_image_estimate[z][y][x] - current_image_estimate[z+dz][y+dy][x+dx]);
+              elemT current  = weights[dz][dy][dx] * (1/(this->scalar))* surrogate(this->scalar*temp);
+
+              if (do_kappa)
+                current *=
+                        (*kappa_ptr)[z][y][x] * (*kappa_ptr)[z+dz][y+dy][x+dx];
+
+              gradient += current;
+            }
+
+        parabolic_surrogate_curvature[z][y][x]= gradient * this->penalisation_factor;
+      }
+    }
+  }
+
+  info(boost::format("parabolic_surrogate_curvature max %1%, min %2%\n") % parabolic_surrogate_curvature.find_max() % parabolic_surrogate_curvature.find_min());
+}
+
+template <typename elemT>
+Succeeded
+LogcoshPrior<elemT>::
+accumulate_Hessian_times_input(DiscretisedDensity<3,elemT>& output,
+                               const DiscretisedDensity<3,elemT>& current_estimate,
+                               const DiscretisedDensity<3,elemT>& input) const
+{
+  // TODO this function overlaps enormously with parabolic_surrogate_curvature
+  // the only difference is that parabolic_surrogate_curvature uses input==1
+
+  assert( output.has_same_characteristics(input));
+  if (this->penalisation_factor==0)
+  {
+    return Succeeded::yes;
+  }
+
+  DiscretisedDensityOnCartesianGrid<3,elemT>& output_cast =
+          dynamic_cast<DiscretisedDensityOnCartesianGrid<3,elemT> &>(output);
+
+  if (weights.get_length() ==0)
+  {
+    compute_weights(weights, output_cast.get_grid_spacing(), this->only_2D);
+  }
+
+  const bool do_kappa = !is_null_ptr(kappa_ptr);
+
+  if (do_kappa && !kappa_ptr->has_same_characteristics(input))
+    error("LogcoshPrior: kappa image has not the same index range as the reconstructed image\n");
+
+  const int min_z = output.get_min_index();
+  const int max_z = output.get_max_index();
+  for (int z=min_z; z<=max_z; z++)
+  {
+    const int min_dz = max(weights.get_min_index(), min_z-z);
+    const int max_dz = min(weights.get_max_index(), max_z-z);
+
+    const int min_y = output[z].get_min_index();
+    const int max_y = output[z].get_max_index();
+
+    for (int y=min_y;y<= max_y;y++)
+    {
+      const int min_dy = max(weights[0].get_min_index(), min_y-y);
+      const int max_dy = min(weights[0].get_max_index(), max_y-y);
+
+      const int min_x = output[z][y].get_min_index();
+      const int max_x = output[z][y].get_max_index();
+
+      for (int x=min_x;x<= max_x;x++)
+      {
+        const int min_dx = max(weights[0][0].get_min_index(), min_x-x);
+        const int max_dx = min(weights[0][0].get_max_index(), max_x-x);
+
+        elemT result = 0;
+        for (int dz=min_dz;dz<=max_dz;++dz)
+          for (int dy=min_dy;dy<=max_dy;++dy)
+            for (int dx=min_dx;dx<=max_dx;++dx)
+            {
+              elemT temp = input[z][y][x] - input[z+dz][y+dy][x+dx];
+              elemT current = weights[dz][dy][dx] * surrogate(temp) * input[z+dz][y+dy][x+dx];
```

Comment:
In `accumulate_Hessian_times_input()`: This is not the hessian. This is the `parabolic_surrogate_curvature_times_input`